classes/processor/processor_lowercase.php

```diff
@@ -0,0 +1,334 @@
+
+        if (!empty($this->get_settings()['csvfields'])) {
+            $this->csvfields = explode('||', $this->get_settings()['csvfields']);
+        }
+
+        $this->now = date('YmdHis', time());
+        $this->filedir = $CFG->dataroot . DIRECTORY_SEPARATOR . $this->get_short_name();
+
+        check_dir_exists($this->filedir);
+    }
+
+    /**
+     * @inheritdoc
+     *
+     * - csvfields: a list of csv field that need to be updated. The names should be separated using ||.
+     *   E.g. username||email||firstname
+     *
+     * - csvdelimiter: a delimiter used in the csv file.
+     *
+     */
+    public function create_config_form_elements(\MoodleQuickForm $mform) {
+        $elements = parent::create_config_form_elements($mform);
+
+        $fields = array(
+            'csvfields' => new config_field(
+                'csvfields',
+                'Fields to modify',
+                'textarea',
+                $this->settings['csvfields'],
+                PARAM_ALPHAEXT
+            ),
+            '' => new config_field(
+                'csvdelimiter',
+                'Csv fields delimiter',
+                'text',
+                $this->settings['csvdelimiter'],
+                PARAM_ALPHAEXT
+            ),
+        );
+
+        return array_merge($elements, $this->get_config_form_elements($mform, $fields));
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function validate_config_form_elements($data, $files, $errors) {
+        $errors = parent::validate_config_form_elements($data, $files, $errors);
+
+        if (!empty($data[$this->get_config_form_prefix() . 'csvfields'])) {
+
+            if (empty($data[$this->get_config_form_prefix() . 'csvdelimiter'])) {
+                $errors[$this->get_config_form_prefix() . 'csvdelimiter'] = 'Csv fields delimiter key could not be empty';
+            } else {
+                if (!array_key_exists($data[$this->get_config_form_prefix() . 'csvdelimiter'], $this->get_delimiter_options())) {
+                    $errors[$this->get_config_form_prefix() . 'csvdelimiter'] = 'Invalid Csv fields delimiter';
+                }
+            }
+        }
+
+        return $errors;
+    }
+
+    /**
+     * Return delimiter option list for config settings of the processor.
+     *
+     * @return array
+     */
+    protected function get_delimiter_options() {
+        return array(
+            ',' => 'Comma (,)',
+            ';' => 'Semi-colon (;)',
+            ':' => 'Colon (:)',
+            '\t' => 'Tab (\t)',
+            '|' => 'Pipe (|)',
+        );
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function process() {
+        parent::process();
+
+        $result = $this->source->extract();
+
+        if (!empty($result->get_supported_formats())) {
+            $files = $result->get_data('files');
+
+            if (!empty($files)) {
+                $newfiles = array();
+
+                foreach ($files as $file) {
+                    try {
```

Comment:
Can the bulk of what happens inside this loop be extracted out into it's own method? Some thing like process_file().
There is a bit much going on here to leave in one big loop.